### ISSUE TITLE
If I'm using the kvm-qemu.sh do I need to do further patching? Currently rebuilding CAPE. [Question]

### ISSUE THREAD

**Issue by powerade661** [kevoreilly/CAPEv2#273]
## This is opensource and you getting __free__ support so be friendly!

Title says it all, if I have installed QEMU with the script developed by D00m3drav3n, do I need to patch anything besides the SATA to IDE within virt-manager? The guide says antivm, does this apply to me if I used the script? 

**From the blogpost**

![image](https://user-images.githubusercontent.com/10760510/93149041-f3089f80-f6ba-11ea-9eab-8ca2d94ae291.png)

Thank you for your help.

**Comment by doomedraven** [comment]
You need to patch vm definition aka vm xml file via virt-manager or virsh
edit X

**Comment by powerade661** [comment]
Alright, so I followed the guide, but I am getting the following from Pafish. Any ideas? :(

**Current CPU configuration**
![image](https://user-images.githubusercontent.com/10760510/93264341-98c81700-f76c-11ea-899a-93d85151276b.png)

**From the blogspot**

![image](https://user-images.githubusercontent.com/10760510/93264386-a8476000-f76c-11ea-84e9-6589a08bcde6.png)

**Alert from pafish**
![image](https://user-images.githubusercontent.com/10760510/93263560-8a2d3000-f76b-11ea-9170-6f280251e8f0.png)

**Comment by doomedraven** [comment]
First detection is that you have something bad, second is ok and bypassed
in real life

**Comment by powerade661** [comment]
Anyway to bypass or do I need to try and rebuild again?

**Comment by doomedraven** [comment]
You dont need to rebuild, just edit vm xml

**Comment by kevoreilly** [comment]
Hi powerade661, try running (or re-processing) Pafish with the rdtsc evasion sig I just pushed:

https://github.com/kevoreilly/CAPEv2/commit/47705d52e5e7c2f9c23b28e5c9b21b38ee1eeed3

Let me know if that works

**Comment by powerade661** [comment]
Hi, Kevin I will run this and get back with you. This looks like it sets the zero flag for the VM_exit instruction only if I'm reading this right? It does not mitigate the rdtsc cpu timestamp counters? That's the issue I am having and I'm not really sure where to edit it to evade it. I followed the guide to the letter, unchecked copy host config, specified my own CPU and it's still not working, I read somewhere that +rdtscp is a switch that one can use to possibly turn it off, but I'm not that we'll versed in evasion for Qemu only VirtualBox which is what I built it with to begin with.

**Comment by doomedraven** [comment]
did you follow https://www.doomedraven.com/2016/#modifying-kvm-qemu-kvm-settings-for-malware-analysis ?

**Comment by powerade661** [comment]
Yep. That's what has me puzzled, there isn't anything about that in there, only editing your MAC address, modifying the CPU to erase the cpuid bit and changing the bios.(all which work great) but couldn't find anything about rdtsc cpu timestamp counters.

**Comment by doomedraven** [comment]
well i don't use nothing else, and uniq detection is rtdsc vm-exit, the first one is defeated, so idk what is wrong on your side sorry

**Comment by powerade661** [comment]
So I have a theory, I'm not sure if this could be causing it? My VM actually froze and I had to restart it and in reset it. Could that be throwing things off?

**Comment by doomedraven** [comment]
shouldn't, as that is more low leven than just running vm, in your xml file you just need OS, bios and cpu/kvm=off, nothing else from extras

**Comment by powerade661** [comment]
Could it be the paravirtualization that I'm using? Minimal

**Comment by doomedraven** [comment]
so im correct, you are on vbox not kvm?

**Comment by powerade661** [comment]
I am running vbox and inside of that KVM

**Comment by doomedraven** [comment]
well thats is your problem i guess, we don't use vbox at all, soo sorry can't help here

**Comment by powerade661** [comment]
I fixed it. 

I executed the following on my VM for anyone who is using vbox.

Find the path to vboxmanage in your virtualbox or oracle folder.

Then execute the following commands

vboxmanage setextradata "VM NAME HERE"  VBoxInternal/TM/TSCMode RealTSCOffset
vboxmanage setextradata "VM NAME HERE"  VBoxInternal/CPUM/SSE4.1 1
vboxmanage setextradata "VM NAME HERE"  VBoxInternal/CPUM/SSE4.2 1

**Comment by kevoreilly** [comment]
Hi powerade661 - yes that Pafish sig is just for Pafish when you have a hypervisor that is detected by rdtsc.

If you have achieved no detection within a vbox vm nested in a kvm vm, then I am slightly amazed... but still horrified at the setup! We definitely don't recommend nested vms but glad you got it working.

**Comment by kevoreilly** [comment]
There is also fake-rdtsc option in the debugger but that's another story...

**Comment by powerade661** [comment]
Here is the evidence of that. 

**NOTE: this is not running the CAPE agent**

![image](https://user-images.githubusercontent.com/10760510/93338489-02d5d000-f7f0-11ea-8f53-f943692db2c1.png)

**Comment by kevoreilly** [comment]
Aha so the latter one is still detected? This is the one targeted by the sig

**Comment by powerade661** [comment]
And after waiting, here is my final result, again without the CAPE agent

![image](https://user-images.githubusercontent.com/10760510/93339866-b3909f00-f7f1-11ea-8d6d-c0bd708a147e.png)

**Comment by kevoreilly** [comment]
I will be interested to see if you swat that final red with capemon

**Comment by powerade661** [comment]
I have no doubt!

**Comment by powerade661** [comment]
So it still didn't remove the final red unfortunately. 
![image](https://user-images.githubusercontent.com/10760510/93345457-74198100-f7f8-11ea-8e65-a897476858b4.png)

**Comment by doomedraven** [comment]
yep nothing will remove it

**Comment by powerade661** [comment]
But you said it was bypassed, right doomed? I wonder if I provide it the fake rtdtsc if it bypass it.

**Comment by doomedraven** [comment]
`rtdst` but not `rtdst vm exit`, that is different thing

**Comment by powerade661** [comment]
Oh I just realized I never installed pywintrace and only the agent, let me run it again and get back to you.

**Comment by doomedraven** [comment]
just in case https://www.qemu.org/2019/11/15/micro-optimizing-kvm-vmexits/

**Comment by kevoreilly** [comment]
I have just experimented with tracing the entire execution of pafish in the debugger whilst using fake-rdtsc - the current hard-coded time delta is too big for pafish (I set it when experimenting with Guloader) so I've made the delta configurable. It's then possible to trace the entire execution silently and passing all checks. I'll publish a new monitor with configurable delta (e.g. fake-rdtsc=5000 or whatever) soon so watch this space!
